File: desktop/src/lib/shared/services/catalog/GutendexDataSource.ts
/**
 * Gutendex datasource — metadata/download authority (`copyright=false`).
 * Plain `fetch` + identified UA; `connect-src` allowlisted in tauri.conf.json.
 */
import { catalogError, isCatalogError } from './errors';
import { DESKTOP_USER_AGENT, clampPageSize, fetchWithRetry, toCatalogError } from './policy';
import { mapGutendexBook, type GutendexRecord } from './mappers';
import type { CatalogBook } from './CatalogProvider';

export const GUTENDEX_BASE_URL = 'https://gutendex.com';

export interface GutendexSearchResponse {
  count: number;
  results: GutendexRecord[];
}

export class GutendexDataSource {
  constructor(
    private readonly fetchFn: typeof fetch = fetch,
    private readonly userAgent: string = DESKTOP_USER_AGENT,
  ) {}

  private async getJson(path: string): Promise<unknown> {
    try {
      const res = await fetchWithRetry(
        `${GUTENDEX_BASE_URL}${path}`,
        { headers: { 'User-Agent': this.userAgent, Accept: 'application/json' } },
        this.fetchFn,
      );
      return await res.json();
    } catch (err) {
      if (isCatalogError(err)) throw err;
      toCatalogError(err, 'gutendex request failed');
    }
  }

  /** Search PD books; in-copyright records are excluded by the mapper. */
  async search(
    query: string,
    page: number,
    pageSize = 24,
  ): Promise<{
    books: CatalogBook[];
    totalCount: number;
  }> {
    const size = clampPageSize(pageSize);
    const params = new URLSearchParams({ search: query, page: String(page) });
    const data = (await this.getJson(`/books/?${params.toString()}`)) as GutendexSearchResponse;
    const books = (data.results ?? [])
      .map(mapGutendexBook)
      .filter((b): b is CatalogBook => b !== null)
      .slice(0, size);
    return { books, totalCount: typeof data.count === 'number' ? data.count : books.length };
  }

  /** Fetch one book by numeric id; unknown ids surface NOT_FOUND. */
  async getById(numericId: number): Promise<CatalogBook> {
    const data = (await this.getJson(`/books/${numericId}/`)) as GutendexRecord;
    const book = mapGutendexBook(data);
    if (!book) throw catalogError('NOT_FOUND', `gutendex book ${numericId} unavailable`);
    return book;
  }
}

File: android/app/src/main/java/com/nextpage/data/remote/catalog/GutendexDataSource.kt
package com.nextpage.data.remote.catalog

import java.net.URLEncoder
import kotlinx.serialization.json.Json

const val GUTENDEX_BASE_URL = "https://gutendex.com"

/**
 * Gutendex datasource — metadata/download authority (`copyright=false`).
 * Ktor transport + identified UA; PD filtering happens in the mapper.
 */
open class GutendexDataSource(
    private val transport: CatalogHttpTransport,
    private val baseUrl: String = GUTENDEX_BASE_URL,
    private val json: Json = Json { ignoreUnknownKeys = true }
) {

    /** Search PD books; in-copyright records are excluded by the mapper. */
    open suspend fun search(
        query: String,
        page: Int,
        pageSize: Int = DEFAULT_PAGE_SIZE
    ): CatalogSearchResult {
        val size = clampPageSize(pageSize)
        val params = "search=${query.encode()}&page=$page"
        val res = transport.getWithRetry("$baseUrl/books/?$params")
        val data = json.decodeFromString<GutendexSearchResponse>(res.body)
        val books = data.results.mapNotNull(::mapGutendexBook).take(size)
        return CatalogSearchResult(books, data.count ?: books.size)
    }

    /** Fetch one book by numeric id; unknown ids surface NOT_FOUND. */
    open suspend fun getById(numericId: Int): CatalogBook {
        val res = transport.getWithRetry("$baseUrl/books/$numericId/")
        val record = json.decodeFromString<GutendexRecord>(res.body)
        return mapGutendexBook(record)
            ?: throw catalogError(CatalogErrorCode.NOT_FOUND, "gutendex book $numericId unavailable")
    }

    private fun String.encode(): String = URLEncoder.encode(this, Charsets.UTF_8.name())
}
